rust/src/treefile.rs

```diff
@@ -2227,8 +2227,14 @@ pub(crate) fn treefile_new(
 }
 
 /// Create a new treefile from a string.
-pub(crate) fn treefile_new_from_string(buf: &str) -> CxxResult<Box<Treefile>> {
-    Ok(Treefile::new_from_string(buf).context("Parsing treefile from string")?)
+// Make `client` into an enum?
+pub(crate) fn treefile_new_from_string(buf: &str, client: bool) -> CxxResult<Box<Treefile>> {
```

Comment:
Maybe a cleaner way to structure this actually is to have fully separate `DerivationTreeComposeConfig` and `BaseTreeComposeConfig` structs. And then `Treefile` holds an enum which is either one of the two. Though hmm, we'd have a `CommonComposeConfig` struct holding common fields flattened into both, which is a little awkward (Golang's promoted fields would've been nice for this).

Reply:
https://stackoverflow.com/questions/32552593/is-it-possible-for-one-struct-to-extend-an-existing-struct-keeping-all-the-fiel mentions (ab)using `Deref` for this which could work out in this case.